test/box-luatest/fully-temporary_spaces_test.lua

```diff
@@ -520,3 +520,40 @@ g.test_meta_data_not_persisted = function()
 
     check_wal_and_snap({})
 end
+
+-- check memory is not leaked when dropping temporary spaces
+g.test_no_memory_leak_after_drop = function()
+    g.server:exec(function()
+        -- Create a `weak_refs` table, with a weak reference to `object`
+        local object = {'object'}
+        local weak_refs = setmetatable({object}, {__mode = 'v'})
+
+        local s = box.schema.space.create('temp', { type = 'temporary' })
+
+        -- Create a strong reference to the same `object` which belongs to
+        -- the on_replace trigger of the temporary space.
+        s:on_replace((function()
+            local strong_ref = object
+            return function() print(strong_ref) end
+        end)())
+
+        -- Delete the original strong reference to `object`
+        object = nil
+
+        -- Trigger garbage collection (needs to be called twice to collect
+        -- memory freed by __gc methods invoked on the first call).
+        collectgarbage()
```

Comment:
Nit: the comment and an extra call to `collectgarbage()` are only needed in the second case, when you drop the space. The "double collectgarbage()" thing doesn't apply to normal lua objects.

Reply:
can something bad happen because of the additional `collectgarbage()` call?

Reply:
No, but it's confusing. And the comment describes a situation which's not possible in this place. The `object` doesn't have a `__gc` metamethod (and can't possibly have one AFAICS). 

The comment and double `collectgarbage` both belong to the second place. Altough even there they seem unnecessary. The test continues working 100% the same when I remove the extra `collectgarbage()` calls:
```diff
diff --git a/test/box-luatest/fully-temporary_spaces_test.lua b/test/box-luatest/fully-temporary_spaces_test.lua
index 783fd42f2f..e7f75bfc71 100644
--- a/test/box-luatest/fully-temporary_spaces_test.lua
+++ b/test/box-luatest/fully-temporary_spaces_test.lua
@@ -543,7 +543,6 @@ g.test_no_memory_leak_after_drop = function()
         -- Trigger garbage collection (needs to be called twice to collect
         -- memory freed by __gc methods invoked on the first call).
         collectgarbage()
-        collectgarbage()
 
         -- on_replace trigger is still live, so `object` still exists.
         t.assert(weak_refs[1])
@@ -551,7 +550,6 @@ g.test_no_memory_leak_after_drop = function()
         -- space.drop should free the on_replace triggers
         s:drop()
         collectgarbage()
-        collectgarbage()
 
         -- on_replace trigger was destroyed and so was `object`
         t.assert(not weak_refs[1])
```

Reply:
The point of the double `collectgarbage` call is that in Lua it is often not enough to just call it once because after the first call the memory may become unused but not yet freed (for the reason explained in the comment). And because the test is checking for memory leaks, it's trying to do a best effort at freeing that memory.

I agree that currently these calls are redundant and removing them doesn't change anything, but I do believe it is possible that at some point in the future something may change and these calls will become necessary. So if there's nothing bad that can happen because of the double collectgarbage calls, I'd rather leave them in the test.

Reply:
Ok, let's keep it your way.